helpers/MigrationHelper.ts

```diff
@@ -0,0 +1,246 @@
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function log(target: any, key: string, descriptor: PropertyDescriptor) {
+  const originalMethod = descriptor.value;
+
+  descriptor.value = async function (...args: any[]) {
+    const start = Date.now();
+    console.log(`========= Executing function ${key}...==========`);
+    const result = await originalMethod.apply(this, args);
+    const end = Date.now();
+    const time = (end - start) / 1000;
+    console.log(
+      `========= Finished executing function ${key}, duration: ${time} seconds =========`
+    );
+    return result;
+  };
+
+  return descriptor;
+}
+
+export interface EventLogEntry {
+  blockNumber: number;
+  blockHash: string;
+  transactionIndex: number;
+  removed: boolean;
+  address: string;
+  data: string;
+  topics: string[];
+  transactionHash: string;
+  logIndex: number;
+  event: string;
+  eventSignature: string;
+  args: {
+    type: string;
+    hex: string;
+  }[];
+}
+
+export interface MigrationResult {
+  migratedData: {
+    args: {
+      type: string;
+      hex: string;
+    }[];
+    tx: string;
+  }[];
+  error: string | null;
+  index: number;
+  receipt: unknown;
+}
+
+export interface IContractMigrationSteps {
+  getInitContract(contractMeta: {
+    address?: string;
+    contractName?: string;
+    abi?: ContractInterface;
+    bytecode?: string | BytesLike;
+  }): Promise<Contract>;
+
+  populateData(contract: Contract, stateTransitionPayload: any[]): Promise<void>;
+
+  readEventLogData(
+    contract: Contract,
+    firstEventBlock: number, //29831814
+    eventsChunkSize: number,
+    eventName?: string,
+    fileName?: string
+  ): Promise<EventLogEntry[]>;
+
+  migrateData(
+    data: EventLogEntry[],
+    populateDataFn: (...args) => Promise<any>,
+    fileName?: string
+  ): Promise<MigrationResult>;
+
+  upgradeContract(contract: Contract): Promise<any>;
+}
+
+export abstract class ContractMigrationSteps implements IContractMigrationSteps {
+  constructor(protected readonly _signer: SignerWithAddress | Wallet) {}
+
+  abstract populateData(contract: Contract, stateTransitionPayload: any[]): Promise<void>;
+
+  abstract upgradeContract(contract: Contract, afterUpgrade?: () => Promise<void>): Promise<any>;
+
+  @log
+  getInitContract(contractMeta: {
+    contractNameOrAbi?: string | any[];
+    address?: string;
+  }): Promise<Contract> {
+    if (Object.keys(contractMeta).every((key) => !contractMeta[key])) {
+      throw new Error("contract meta is empty");
+    }
+
+    if (contractMeta.address && contractMeta.contractNameOrAbi) {
+      return ethers.getContractAt(
+        contractMeta.contractNameOrAbi,
+        contractMeta.address,
+        this._signer
+      );
+    }
+
+    throw new Error("Invalid contract meta");
+  }
+
+  @log
+  async readEventLogData(
+    contract: Contract,
+    firstEventBlock: number,
+    eventsChunkSize: number,
+    eventName = "StateUpdated",
```

Comment:
Is it worth hardcoding default eventName to "StateUpdated" for the abstract class?